components/dashboard/src/data/setup.tsx

```diff
@@ -0,0 +1,86 @@
+import {
+    PersistedClient,
+    Persister,
+    PersistQueryClientProvider,
+    PersistQueryClientProviderProps,
+} from "@tanstack/react-query-persist-client";
+import { QueryClient } from "@tanstack/react-query";
+import { ReactQueryDevtools } from "@tanstack/react-query-devtools";
+import { FunctionComponent } from "react";
+
+// This is used to version the cache
+// If data we cache changes in a non-backwards compatible way, increment this version
+// That will bust any previous cache versions a client may have stored
+const CACHE_VERSION = "1";
+
+export const setupQueryClientProvider = () => {
+    const client = new QueryClient();
+    const queryClientPersister = createIDBPersister();
+
+    const persistOptions: PersistQueryClientProviderProps["persistOptions"] = {
+        persister: queryClientPersister,
+        // This allows the client to persist up to 24 hours
+        // Individual queries may expire prior to this though
+        maxAge: 1000 * 60 * 60 * 24, // 24 hours
+        buster: CACHE_VERSION,
+    };
+
+    // Return a wrapper around PersistQueryClientProvider w/ the query client options we setp
+    const GitpodQueryClientProvider: FunctionComponent = ({ children }) => {
+        return (
+            <PersistQueryClientProvider client={client} persistOptions={persistOptions}>
+                {children}
+                <ReactQueryDevtools initialIsOpen={false} />
+            </PersistQueryClientProvider>
+        );
+    };
+
+    return GitpodQueryClientProvider;
+};
+
+// Persister that uses IndexedDB
+function createIDBPersister(idbValidKey: IDBValidKey = "gitpodQueryClient") {
+    // Track a flag that indicates if we're attempting to persist the client
+    // Some browsers/versions don't support using indexed-db w/ certain settings or in private mode
+    // If we get an error performing an operation, we'll disable persistance and assume it's not supported
+    let persistanceActive = true;
+
+    return {
+        persistClient: async (client: PersistedClient) => {
+            if (!persistanceActive) {
+                return;
+            }
+
+            try {
+                await set(idbValidKey, client);
+            } catch (e) {
+                console.error("unable to persist query client");
+                persistanceActive = false;
+            }
+        },
+        restoreClient: async () => {
+            try {
+                const client = await get<PersistedClient>(idbValidKey);
+                return client;
+            } catch (e) {
+                console.error("unable to load query client from cache");
+                persistanceActive = false;
+            }
+        },
+        removeClient: async () => {
+            try {
+                await del(idbValidKey);
+            } catch (e) {
+                console.error("unable to remove query client");
+                persistanceActive = false;
+            }
+        },
+    } as Persister;
```

Comment:
How come we need to force the type to `Persister` using `as` here?

Reply:
Just to type the object closer to where it's defined, vs. where it's consumed. I honestly didn't think too much about it, as I went with the example react-query has on their docs, and they did it this way 😄 I'm going to shift it to a return type on the function rather than inline like this, seems more intuitive that way than how I have it atm. Thanks!